src/app/dim-ui/WeaponGroupingIcon.tsx

```diff
@@ -0,0 +1,104 @@
+import { TagValue } from '@destinyitemmanager/dim-api-types';
+import TagIcon from 'app/inventory/TagIcon';
+import { AmmoIcon } from 'app/item-popup/ItemPopupHeader';
+import { vaultWeaponGroupingSettingSelector } from 'app/settings/vault-grouping';
+import { VaultGroupIconValue } from 'app/shell/item-comparators';
+import { DestinyAmmunitionType } from 'bungie-api-ts/destiny2';
+import { useSelector } from 'react-redux';
+import ElementIcon from './ElementIcon';
+import { getWeaponTypeSvgIconFromCategoryHashes } from './svgs/itemCategory';
+
+const VALID_AMMO_TYPES: Record<DestinyAmmunitionType, boolean> = {
+  [DestinyAmmunitionType.None]: true,
+  [DestinyAmmunitionType.Primary]: true,
+  [DestinyAmmunitionType.Special]: true,
+  [DestinyAmmunitionType.Heavy]: true,
+  [DestinyAmmunitionType.Unknown]: true,
+};
+
+const VALID_TAGS: Record<TagValue, boolean> = {
+  favorite: true,
+  keep: true,
+  infuse: true,
+  archive: true,
+  junk: true,
+};
+
+export default function WeaponGroupingIcon({
+  iconValue,
+  className,
+}: {
+  iconValue: VaultGroupIconValue;
+  className?: string;
+}) {
+  const vaultWeaponGroupingSetting = useSelector(vaultWeaponGroupingSettingSelector);
+
+  if (!vaultWeaponGroupingSetting || typeof iconValue === 'undefined') {
+    return null;
+  }
+
+  if (
+    vaultWeaponGroupingSetting === 'typeName' &&
+    typeof iconValue === 'object' &&
+    // eslint-disable-next-line no-implicit-coercion
+    !!iconValue &&
```

Comment:
```suggestion
    Boolean(iconValue) &&
```

Reply:
This is a type-guard. `Boolean()` doesn't prove that the thing that's passed in exists, whereas `!!` does.

The alternative is to manually create a type guard for this, but it's less type-safe.

Reply:
I actually used to enforce `Boolean` with linting myself until I got more into TypeScript and then dropped it in favor of double negation.

Reply:
Yeah, that's https://github.com/microsoft/TypeScript/issues/16655. I don't love how implicit/magical `!!` is but I'll consider reversing the lint. Generally we try to structure types such that this isn't an issue, since it's confusing to have to do this type of thing.

Reply:
Also see https://github.com/typescript-eslint/typescript-eslint/issues/6316#issuecomment-1376059324